src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs

```diff
@@ -1317,6 +1348,167 @@ public void Dispose()
+            {
+                var dto = new ReplicationItemDto
+                {
+                    TransactionMarker = item.TransactionMarker,
+                    Type = item.Type,
+                    Id = item.Id,
+                    Position = item.Position,
+                    ChangeVector = item.ChangeVector,
+                    DocumentSize = item.DocumentSize,
+                    Collection = item.Collection,
+                    LastModifiedTicks = item.LastModifiedTicks,
+                    Flags = item.Flags,
+                    Key = item.Key.ToString(),
+                    Base64Hash = item.Base64Hash.ToString()
+                };
+
+                dto.Name = item.Name.Content.HasValue
+                    ? context.GetLazyStringValue(item.Name.Content.Ptr)
+                    : null;
+
+                dto.ContentType = item.Name.Content.HasValue
+                    ? context.GetLazyStringValue(item.ContentType.Content.Ptr)
+                    : null;
+
+                return dto;
+            }
+        }
+    }
+
+    internal class MergedUpdateDatabaseChangeVectorCommandDto : TransactionOperationsMerger.IReplayableCommandDto<IncomingReplicationHandler.MergedUpdateDatabaseChangeVectorCommand>
+    {
+        public string ChangeVector;
+        public long LastDocumentEtag;
+        public string SourceDatabaseId;
+
+        public IncomingReplicationHandler.MergedUpdateDatabaseChangeVectorCommand ToCommand(DocumentsOperationContext context, DocumentDatabase database)
+        {
+            var command = new IncomingReplicationHandler.MergedUpdateDatabaseChangeVectorCommand(ChangeVector, LastDocumentEtag, SourceDatabaseId, new AsyncManualResetEvent());
+            return command;
+        }
+    }
+
+
+    internal class MergedDocumentReplicationCommandDto : TransactionOperationsMerger.IReplayableCommandDto<IncomingReplicationHandler.MergedDocumentReplicationCommand>
+    {
+        public ReplicationItemDto[] ReplicatedItemDtos;
+        public long LastEtag;
+        public string Buffer;
+        public string SourceDatabaseId;
+        public KeyValuePair<string, Stream>[] ReplicatedAttachmentStreams;
+
+        public IncomingReplicationHandler.MergedDocumentReplicationCommand ToCommand(DocumentsOperationContext context, DocumentDatabase database)
+        {
+            var connectionInfo = new IncomingConnectionInfo
+            {
+                SourceDatabaseId = SourceDatabaseId
+            };
+            var replicationItems = ReplicatedItemDtos.Select(d => d.ToItem(context)).ToList();
+
+            var replicatedAttachmentStreams = ReplicatedAttachmentStreams.Select(i => CreateReplicationAttachmentStream(context, i)).ToArray();
+            var replicationHandler = new IncomingReplicationHandler(database, replicationItems, connectionInfo, database.ReplicationLoader, replicatedAttachmentStreams);
```

Comment:
Can we avoid doing that?

Reply:
Avoiding what? Linq?

Reply:
@haludi I think @ppekrol was referencing the creationg of IncomingReplicationHandler, which is pretty heavy object, that does a lot of stuff under the hood. I think that at this stage it will be a bit hard to get rid of it, because of the none trivial code in MergedDocumentReplicationCommand.Execute. But yes, it looks like a big TODO

Reply:
An issue created
[RavenDB-11857](http://issues.hibernatingrhinos.com/issue/RavenDB-11857)